engine/consensus/approvals/core.go

```diff
@@ -0,0 +1,392 @@
+	// Same functor is used by AssignmentCollector to reuse cache instead of using state for every
+	// height lookup.
+	core.getCachedBlockHeight = func(blockID flow.Identifier) (uint64, error) {
+		height, cached := core.blockHeightLookupCache.Peek(blockID)
+		if !cached {
+			// check if we already have the block the approval pertains to
+			head, err := core.state.AtBlockID(blockID).Head()
+			if err != nil {
+				if !errors.Is(err, storage.ErrNotFound) {
+					return 0, fmt.Errorf("failed to retrieve header for block %x: %w", blockID, err)
+				}
+				return 0, engine.NewUnverifiableInputError("no header for block: %v", blockID)
+			}
+
+			height = head.Height
+			core.blockHeightLookupCache.Add(blockID, height)
+		}
+		return height.(uint64), nil
+	}
+
+	return core
+}
+
+func (c *approvalProcessingCore) lastSealedHeight() uint64 {
+	return atomic.LoadUint64(&c.atomicLastSealedHeight)
+}
+
+func (c *approvalProcessingCore) lastFinalizedHeight() uint64 {
+	return atomic.LoadUint64(&c.atomicLastFinalizedHeight)
+}
+
+// WARNING: this function is implemented in a way that we expect blocks strictly in parent-child order
+// Caller has to ensure that it doesn't feed blocks that were already processed or in wrong order.
+func (c *approvalProcessingCore) OnFinalizedBlock(finalizedBlockID flow.Identifier) {
+	finalized, err := c.state.AtBlockID(finalizedBlockID).Head()
+	if err != nil {
+		c.log.Fatal().Err(err).Msgf("could not retrieve header for finalized block %s", finalizedBlockID)
+	}
+
+	payload, err := c.payloads.ByBlockID(finalizedBlockID)
+	if err != nil {
+		c.log.Fatal().Err(err).Msgf("could not retrieve payload for finalized block %s", finalizedBlockID)
+	}
+
+	// it's important to use atomic operation to make sure that we have correct ordering
+	atomic.StoreUint64(&c.atomicLastFinalizedHeight, finalized.Height)
+
+	lastSealed, err := c.state.Sealed().Head()
+	if err != nil {
+		c.log.Fatal().Err(err).Msgf("could not retrieve last sealed block")
+	}
+
+	// it's important to use atomic operation to make sure that we have correct ordering
+	atomic.StoreUint64(&c.atomicLastSealedHeight, lastSealed.Height)
+
+	sealsCount := len(payload.Seals)
+	sealedResultIds := make([]flow.Identifier, sealsCount)
+	for i, seal := range payload.Seals {
+		sealedResultIds[i] = seal.ResultID
+	}
+
+	// cleanup collectors for already sealed results
+	c.eraseCollectors(sealedResultIds)
+	collectors := c.allCollectors()
+
+	// check if there are stale results qualified for emergency sealing
+	err = c.checkEmergencySealing(collectors, finalized.Height)
+	if err != nil {
+		c.log.Err(err).Msgf("could not check emergency sealing at block %v", finalizedBlockID)
+	}
+
+	collectors = c.cleanupStaleCollectors(collectors, lastSealed.Height)
+	// to those collectors that are not stale, report finalization event to cleanup orphan blocks
+	for _, collector := range collectors {
+		collector.OnBlockFinalizedAtHeight(finalizedBlockID, finalized.Height)
+	}
```

Comment:
I'm a bit concerned about these cleanup calls.

The for-loops in them creates an amplify factor as each finalized block will trigger the clean up, which will iterate through the entire `collectors` map. If the iteration is slow, then the `OnFinalizedBlock` will be blocked and further blocking Hotstuff.

I think we don't need to do these cleanup for every finalized block. Instead, we just need to cleanup once a while from the last cleaned up height. 

The cleanup could have delay, as long as it's not blocking the finalization.

So we might need to maintain two state values: `lastSealedHeight` and `lastCleanedupHeight`, so that when a block is finalized, we update the `lastSealedHeight`, and then trigger a call to cleanup, but that call will abort if the previous cleanup has not finished yet. We can implement that with an atomic.Bool value, just like we ensure only one thread is calling `checkSealing`.

If cleanup has finished, we update the `lastCleanedupHeight`, and the next cleanup call will start from the `lastCleanedupHeight` up to the `lastSealedHeight`.

So it's OK that the cleanup falling behind sometimes, but eventually it should catch up. And during the catch up period, it should have no impact to the finalization.

Reply:
I am expecting it to have 0 impact on hotstuff since it `engine` will dedicate where this function will run and most likely it will be run on separate worker. If you still think that we shouldn't cleanup that often then sure, I can do it, but I don't think it's very bad, we always can fix this later if it becomes a problem